include/libp2p/crypto/x25519_provider.hpp

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_INCLUDE_LIBP2P_CRYPTO_X25519_PROVIDER_HPP
+#define LIBP2P_INCLUDE_LIBP2P_CRYPTO_X25519_PROVIDER_HPP
+
+#include <array>
+#include <vector>
+
+#include <gsl/span>
+#include <libp2p/outcome/outcome.hpp>
+
+namespace libp2p::crypto::x25519 {
+
+  using PrivateKey = std::array<uint8_t, 32u>;
+  using PublicKey = std::array<uint8_t, 32u>;
+  struct Keypair {
+    PrivateKey private_key;
+    PublicKey public_key;
+  };
+
+  /**
+   * Diffie Hellman key agreement calculator
+   */
+  class X25519Provider {
+   public:
+    virtual ~X25519Provider() = default;
+
+    /**
+     * Generates a keypair using x25519 curve
```

Comment:
As a matter of terminology: AFAIK x25519 is the DH function, where the name of the underlying curve is Curve25519.